servers/src/main/java/tachyon/master/lineage/LineageMaster.java

```diff
@@ -0,0 +1,415 @@
+  }
+
+  @Override
+  public String getServiceName() {
+    return Constants.LINEAGE_MASTER_SERVICE_NAME;
+  }
+
+  @Override
+  public void processJournalEntry(JournalEntry entry) throws IOException {
+    if (entry instanceof LineageEntry) {
+      mLineageStore.addLineageFromJournal((LineageEntry) entry);
+    } else if (entry instanceof LineageIdGeneratorEntry) {
+      mLineageIdGenerator.fromJournalEntry((LineageIdGeneratorEntry) entry);
+    } else if (entry instanceof AsyncCompleteFileEntry) {
+      asyncCompleteFileFromEntry((AsyncCompleteFileEntry) entry);
+    } else if (entry instanceof PersistFilesEntry) {
+      persistFilesFromEntry((PersistFilesEntry) entry);
+    } else if (entry instanceof RequestFilePersistenceEntry) {
+      requestFilePersistenceFromEntry((RequestFilePersistenceEntry) entry);
+    } else {
+      throw new IOException(ExceptionMessage.UNEXPECETD_JOURNAL_ENTRY.getMessage(entry));
+    }
+  }
+
+  @Override
+  public void start(boolean isLeader) throws IOException {
+    super.start(isLeader);
+    if (isLeader) {
+      mCheckpointExecutionService =
+          getExecutorService().submit(new HeartbeatThread("Checkpoint planning service",
+              new CheckpointPlanningExecutor(mTachyonConf, this),
+              mTachyonConf.getInt(Constants.MASTER_LINEAGE_CHECKPOINT_INTERVAL_MS)));
+      mRecomputeExecutionService =
+          getExecutorService()
+              .submit(new HeartbeatThread("Recomupte service",
+                  new RecomputeExecutor(new RecomputePlanner(mLineageStore, mFileSystemMaster),
+                      mFileSystemMaster),
+                  mTachyonConf.getInt(Constants.MASTER_LINEAGE_RECOMPUTE_INTERVAL_MS)));
+    }
+  }
+
+  @Override
+  public void stop() throws IOException {
+    super.stop();
+    if (mCheckpointExecutionService != null) {
+      mCheckpointExecutionService.cancel(true);
+    }
+    if (mRecomputeExecutionService != null) {
+      mRecomputeExecutionService.cancel(true);
+    }
+  }
+
+  @Override
+  public void streamToJournalCheckpoint(JournalOutputStream outputStream) throws IOException {
+    mLineageStore.streamToJournalCheckpoint(outputStream);
+  }
+
+  public LineageStoreView getLineageStoreView() {
+    return new LineageStoreView(mLineageStore);
+  }
+
+  public long createLineage(List<TachyonURI> inputFiles, List<TachyonURI> outputFiles, Job job)
+      throws InvalidPathException, FileAlreadyExistException, BlockInfoException {
+    // validate input files exist
+    List<TachyonFile> inputTachyonFiles = Lists.newArrayList();
+    for (TachyonURI inputFile : inputFiles) {
+      long fileId;
+      fileId = mFileSystemMaster.getFileId(inputFile);
+      inputTachyonFiles.add(new TachyonFile(fileId));
+    }
+    // create output files
+    List<LineageFile> outputTachyonFiles = Lists.newArrayList();
+    for (TachyonURI outputFile : outputFiles) {
+      long fileId;
+      fileId = mFileSystemMaster.createFile(outputFile, 1, true);
```

Comment:
Also, it seems to me that once some one calls `createLineage` with specified output files (although they don't exist yet), these output files will show up in Tachyon file system with 1 byte each. This concerns me because I feel those files should be different from normal TachyonFiles before they were really generated by the commands.